main/src/ca/uwaterloo/flix/language/phase/Monomorpher.scala

```diff
@@ -813,6 +782,62 @@ object Monomorpher {
     }
   }
 
+  /** Reduces the given associated type and crashes if it is not possible. */
+  private def infallibleReduceAssocType(cst: Ast.AssocTypeConstructor, arg: Type, eqEnv: ListMap[Symbol.AssocTypeSym, Ast.AssocTypeDef])(implicit flix: Flix): Type = {
+    EqualityEnvironment.reduceAssocType(cst, arg, eqEnv) match {
+      case Ok(t) => t
+      case Err(_) => throw InternalCompilerException("Unexpected associated type reduction failure", arg.loc)
+    }
+  }
+
+  /**
+    * Removes [[Type.Alias]] and [[Type.AssocType]], or crashes if some [[Type.AssocType]] is not
+    * reducible.
+    */
+  private def simplify(tpe: Type, eqEnv: ListMap[Symbol.AssocTypeSym, Ast.AssocTypeDef])(implicit flix: Flix): Type = tpe match {
+    case v@Type.Var(_, _) => v
+    case c@Type.Cst(_, _) => c
+    case app@Type.Apply(_, _, _) => normalizeApply(simplify(_, eqEnv), app)
+    case Type.Alias(_, _, tpe, _) => simplify(tpe, eqEnv)
+    case Type.AssocType(cst, arg0, _, _) => simplify(infallibleReduceAssocType(cst, simplify(arg0, eqEnv), eqEnv), eqEnv)
+    case Type.JvmToType(_, loc) => throw InternalCompilerException("unexpected JVM type", loc)
+    case Type.JvmToEff(_, loc) => throw InternalCompilerException("unexpected JVM eff", loc)
+    case Type.UnresolvedJvmType(_, loc) => throw InternalCompilerException("unexpected JVM type", loc)
+  }
+
+  /**
+    * Applies `f` on both sides of the application, then normalizing the remaining type.
+    *
+    * OBS: `f` must not output [[Type.AssocType]] or [[Type.Alias]].
+    */
+  @inline
+  private def normalizeApply(normalize: Type => Type, app: Type.Apply): Type = {
```

Comment:
Can we call it simplify or reduce? It does not normalize because there is no normal form. (But maybe some day)?

Reply:
The function or the argument? `normalizeApply` normalizes types if `normalize` normalizes

Reply:
I guess everywhere-- because there is no normal form.

Reply:
there is a normal form? Not structurally, but monomorpher creates normalized types

Reply:
Wait.. Is there a unique normal form? I.e. a concrete set?

Reply:
Oh right, ordering of effects in the ground union is missing, ill add a ticket

but barring oversights, the intention is that there is a normal form, hence the sorting of rows, etc.